src/node_v8.cc

```diff
@@ -243,6 +249,66 @@ void SetFlagsFromString(const FunctionCallbackInfo<Value>& args) {
   V8::SetFlagsFromString(*flags, static_cast<size_t>(flags.length()));
 }
 
+class JSONOutputStream : public v8::OutputStream {
+  public:
+   JSONOutputStream() {}
+
+   int GetChunkSize() override { return 65536; }
+
+   void EndOfStream() override {}
+
+   WriteResult WriteAsciiChunk(char* data, const int size) override {
+     out_stream_.write(data, size);
+     return kContinue;
+   }
+
+   std::ostringstream& out_stream() { return out_stream_; }
+
+  private:
+   std::ostringstream out_stream_;
+};
+
+void StartCpuProfile(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args[0]->IsString());
+  Isolate* isolate = env->isolate();
+  String::Utf8Value name(isolate, args[0]);
+  Local<String> title = String::NewFromUtf8(isolate,
+                                            *name,
+                                            NewStringType::kNormal,
+                                            name.length())
+                                            .ToLocalChecked();
+  CpuProfilingStatus status = env->cpu_profiler()->StartProfiling(title, true);
+  args.GetReturnValue().Set(Number::New(isolate, static_cast<double>(status)));
+}
+
+void StopCpuProfile(const FunctionCallbackInfo<Value>& args) {
+  Environment* env = Environment::GetCurrent(args);
+  CHECK(args[0]->IsString());
+  Isolate* isolate = env->isolate();
+  String::Utf8Value name(isolate, args[0]);
+  Local<String> title = String::NewFromUtf8(isolate,
+                                            *name,
+                                            NewStringType::kNormal,
+                                            name.length())
+                                            .ToLocalChecked();
+  CpuProfile* profile = env->cpu_profiler()->StopProfiling(title);
+  if (profile) {
+    auto json_out_stream = std::make_unique<JSONOutputStream>();
+    profile->Serialize(json_out_stream.get(),
+                        CpuProfile::SerializationFormat::kJSON);
+    profile->Delete();
+    Local<Value> ret;
+    if (ToV8Value(env->context(), json_out_stream->out_stream().str(), isolate).ToLocal(&ret)) {
+      args.GetReturnValue().Set(ret);
+    } else {
+      args.GetReturnValue().Set(Number::New(isolate, -2));
+    }
+  } else {
+    args.GetReturnValue().Set(Number::New(isolate, -1));
```

Comment:
I'm not a huge fan of polymorphic return values (string in the success case, number in the error case). Instead of returning a value here, perhaps the better option here is to throw an error (e.g. a `THROW_ERR_...(env)` variant. That way the javascript side can be simplified so that it only needs to handle the return value in the success case.